### PR TITLE
Create Job Role Backend

### PR DESCRIPTION
**Pull request by JonnyEvans99**

### CODE REVIEW

**Review by eoinmcbennett — APPROVED** [review]

File: src/main/java/org/kainos/ea/cli/Capability.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package org.kainos.ea.cli;
+
+public class Capability {
+
+    private int capabilityID;
+    private String name;
+
+    public Capability(){
+
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/cli/Capability.java]: don't need empty constructor

File: src/main/java/org/kainos/ea/cli/JobBand.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package org.kainos.ea.cli;
+
+public class JobBand {
+
+    private int jobBandID;
+    private String jobBandName;
+
+    public JobBand(){
+
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/cli/JobBand.java]: don't need empty constructor

File: src/main/java/org/kainos/ea/core/JobValidator.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        //Capability - Need to pass in ID?
+        if(job.getCapabilityID() < 1)
+        {
+            return "CapabilityID is less than 1";
+        }
+        if (job.getCapabilityID()>=1)
+        {
+            return null;
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/core/JobValidator.java]: don't return null here as it won't get to the checks below

File: src/main/java/org/kainos/ea/core/JobValidator.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            return "URL is too long";
+        }
+
+        //Job band
+        if (job.getJobBandID() < 1){
+            return "JobBandID is less than 1";
+        }
+        if (job.getJobBandID()>=1){
+            return null;
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/core/JobValidator.java]: don't return null here as it won't get to the checks below

File: src/test/java/org/kainos/ea/controller/JobControllerUnitTest.java [hunk 5]
```diff
@@ -45,8 +55,34 @@ void getAllJobs_shouldReturn500StatusCodeAndErrorMessage_whenServiceThrowsFailed
...
+        Mockito.when(jobService.createJobRole(jobRequest)).thenReturn(expectedId);
+
+        Response response = jobController.createJobRole(jobRequest);
+
+       assertEquals(response.getStatus(), expectedStatusCode);
+        assertEquals(response.getEntity(), expectedId);
+    }
+
+    @Test
+    void createJobRole_shouldReturn500Response_whenJobRoleServiceThrowsActionFailedException() throws ProjectException, SQLException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/controller/JobControllerUnitTest.java:88]: need another test for ProjectException being thrown from service

File: src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java [hunk 6]
```diff
@@ -35,18 +32,9 @@ void getAllJobs_ShouldReturnListOfJobs() {
...
-            Assertions.assertNotNull(job.getJobURL());
-            Assertions.assertNotNull(job.getJobSpec());
-            Assertions.assertNotNull(job.getCapability());
-            Assertions.assertNotNull(job.getJobBand());
-        }
+        List<Job> allJobs = response.readEntity(List.class);
+
+        Assertions.assertFalse(allJobs.isEmpty());
 
     }
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java:64]: need an integration test for new post endpoint. 1 for successful create and 1 for validation error scenario

File: src/test/java/org/kainos/ea/service/JobBandServiceUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+public class JobBandServiceUnitTest {
+
+    JobBandDao jobBandDao = Mockito.mock(JobBandDao.class);
+    DatabaseConnector databaseConnector = Mockito.mock(DatabaseConnector.class);
+
+    JobBandService jobBandService = new JobBandService(jobBandDao, databaseConnector);
+    Connection connection;
+
+    @Test
+    void getAllJobBands_shouldReturnListOfJobBands_whenDaoReturnsListOfJobBands() throws SQLException, FailedToGetJobBandsException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/JobBandServiceUnitTest.java:32]: need 2 more tests for exception scenarios

File: src/test/java/org/kainos/ea/service/JobServiceUnitTest.java [hunk 8]
```diff
@@ -64,4 +75,33 @@ void getAllJobs_shouldThrowFailedToGetJobsException_whenDatabaseConnectorThrowsS
...
+
+        int result = jobService.createJobRole(jobRequest);
+
+        assertEquals(result, expectedID);
+    }
+
+
+
+    @Test
+    void createJobRole_shouldThrowProjectException_whenValidationFails() throws SQLException, ProjectException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/JobServiceUnitTest.java]: this test isn't checking if validation fails. you should be mocking your validator and checking each possible scenario

**Review by shaunganley — CHANGES_REQUESTED** [review]
missing unit tests for validator

File: src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java [hunk 9]
```diff
@@ -21,14 +32,18 @@ public class JobControllerIntegrationTest {
             new ResourceConfigurationSourceProvider()
     );
 
+    DatabaseConnector databaseConnector = Mockito.mock(DatabaseConnector.class);
+    JobService jobService = Mockito.mock(JobService.class);
+    JobController jobController = new JobController(jobService);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java]: not an integration test if you're mocking the dependencies, this makes it a unit test

File: src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java [hunk 10]
```diff
@@ -38,4 +53,37 @@ void getAllJobs_ShouldReturnListOfJobs() {
...
+    void createJobRole_ShouldCreateJobSuccessfully() throws SQLException, ProjectException {
+        // Assuming you have a JobRequest instance set up
+        JobRequest jobRequest = new JobRequest("Sample Job", 1, "Job spec", "https://example.com", 2);
+
+        // Mock the behavior of the JobService to return a job ID
+        int expectedJobId = 1;
+        Mockito.when(jobService.createJobRole(jobRequest)).thenReturn(expectedJobId);
+
+        // Invoke the controller method
+        Response response = jobController.createJobRole(jobRequest);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java]: not an integration test if you're calling the method, you need to call the URL

Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java]: not an integration test if you're calling the method, you need to call the URL

File: src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java [hunk 11]
```diff
@@ -38,4 +53,37 @@ void getAllJobs_ShouldReturnListOfJobs() {
...
+
+    @Test
+    void createJobRole_ShouldFailWithInvalidData() throws SQLException, ProjectException {
+
+        JobRequest jobRequest = new JobRequest("", 0, "", "", -1);
+        int expectedStatusCode = 400;
+
+        Mockito.doThrow(ProjectException.class).when(jobService).createJobRole(jobRequest);
+
+        Response response = jobController.createJobRole(jobRequest);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobControllerIntegrationTest.java]: not an integration test if you're calling the method, you need to call the URL

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+import org.kainos.ea.core.JobValidator;
+
+import org.mockito.Mockito;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+public class ValidatorUnitTest {
+        @Test
+        public void testValidJobRequest() {
+            JobRequest validJob = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+            JobValidator validator = new JobValidator();
+            assertNull(validator.isValid(validJob));
+        }
+
+    //Negative testing each property
+
+    //JobTitle
+    @Test
+    public void testEmptyJobTitle() {
+        JobRequest emptyJobTitle = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        when(emptyJobTitle.getTitle()).thenReturn(""); // Empty title
+
+        JobValidator validator = new JobValidator();
+        assertEquals("No job title entered", validator.isValid(emptyJobTitle));
+    }
+
+    @Test
+    public void testTooBigJobTitle() {
+        JobRequest jobTitleTooBig = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+        when(jobTitleTooBig.getTitle()).thenReturn("MyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitleMyHugeJobTitle"); // Empty title
+
+        JobValidator validator = new JobValidator();
+        assertEquals("Job title is too long", validator.isValid(jobTitleTooBig));
+    }
+
+    //Capability
+    @Test
+    public void testCapabilityIDNegative() {
+        JobRequest invalidCapabilityID = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        when(invalidCapabilityID.getJobURL()).thenReturn("http://validurl.com");
+        when(invalidCapabilityID.getJobBandID()).thenReturn(2);
+
+        JobValidator validator = new JobValidator();
+        assertEquals("CapabilityID is less than 1", validator.isValid(invalidCapabilityID));
+    }
+
+    @Test
+    public void testCapabilityIDTooBig() {
+        JobRequest invalidCapabilityID = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        when(invalidCapabilityID.getJobBandID()).thenReturn(1);
+
+        JobValidator validator = new JobValidator();
+        assertEquals("CapabilityID is greater than 65,535", validator.isValid(invalidCapabilityID));
+    }
+
+    // Job Spec
+    @Test
+    public void testJobSpecTooLong() {
+        JobRequest jobSpecTooLong = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+        when(jobSpecTooLong.getJobURL()).thenReturn("http://validurl.com");
+        when(jobSpecTooLong.getJobBandID()).thenReturn(2);
+
+        JobValidator validator = new JobValidator();
+        assertEquals("Job spec summary is too long", validator.isValid(jobSpecTooLong));
+    }
+
+    @Test
+    public void testJobSpecEmpty() {
+        JobRequest jobSpecEmpty = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+        when(jobSpecEmpty.getJobBandID()).thenReturn(2);
+
+        JobValidator validator = new JobValidator();
+        assertEquals("Job spec summary is null or less than 0 characters", validator.isValid(jobSpecEmpty));
+    }
+
+    // Job URL
+    @Test
+    public void testURLEmpty() {
+        JobRequest jobURLEmpty = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

File: src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+        when(jobURLEmpty.getJobURL()).thenReturn("");
+        when(jobURLEmpty.getJobBandID()).thenReturn(2);
+
+        JobValidator validator = new JobValidator();
+        assertEquals("No URL entered", validator.isValid(jobURLEmpty));
+    }
+
+    @Test
+    public void testURLTooLong() {
+        JobRequest jobURLTooLong = mock(JobRequest.class);
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/validator/ValidatorUnitTest.java]: you don't need to mock this out, just create an instance of it and pass it in

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — APPROVED** [review]

**Pull request merged**